rfcs/0123-flake-names.md

```diff
@@ -0,0 +1,89 @@
+---
+feature: flake-names
+start-date: 2022-03-12
+author: Anselm Schüler
+co-authors: None
+shepherd-team: None
+shepherd-leader: None
+related-issues: None
+---
+
+# Summary
+[summary]: #summary
+
+Flakes can declare the field `name`.  
+It represents the name of the flake.  
+The derivations for a flake are no longer called `source`, but use the flake name.
+
+# Motivation
+[motivation]: #motivation
+
+Flake-centric workflows often end up with a lot of derivations named “source”, and it’s difficult to navigate this.
+This can also make flake-centric commands friendlier and easier to approach.
+
+# Detailed design
+[design]: #detailed-design
+
+A new supported property for flakes is introduced, `name`.  
+The derivation that contains the flake’s content is called `flake-source-${name}`  
+If a source control tag (e.g. git tag such as V3) is available, it is `flake-source-${name}-${tag}`.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Running `nix flake metadata` on a flake that declares this field displays it at the top.  
+Running `nix flake show` on a flake that declares this field shows the name instead of the URL, followed by the URL in parentheses.
+
+Examples:
+
+File `/example/flake.nix`
+```nix
+{
+  name = "example";
+  outputs = { ... }: {
+    lib.example = "example";
+  };
+}
+```
+
+Shell
+```console
+$ nix eval git+file:///example
+[…] copying flake-source-example
+"example"
+```
+
+Example of interactions:
+
+Shell (using the previous file)
+```
+$ nix flake metadata /example
+Name:          example
+Resolved URL:  git+file:///example
+Locked URL:    …
+$ nix flake show /example
+example (git+file:///home/anselmschueler/Code/example?rev=b0&rev=c714c8624f5d49a9d88e6e24550dd88515923c18)
+└───lib: unknown
+```
+
+# Drawbacks
```

Comment:
The big drawback here (where “big” would probably need to be refined, I’ve no idea how much of an issue it is in practice) is that since the name is part of the store path, two flakes with the same content but a different name will end-up in a different place in the store − which itself means that everything that depends on them would have to be rebuild.

Reply:
But flakes with a different name will necessarily  have different content? Or are the flake.* files not copied to the store?

Reply:
Fair point (I had a convoluted scenario in mind, but that’s probably way out of the scope of this RFC now that I think of it). 

That’s actually highlighting an interesting design challenge: There’s currently a neat separation between the fetching code and the code that knows about what the flake is. `inputs.foo = something` gets translated to a call to `builtins.fetchTree something`, and its result is then fed to the `outputs` function. And the `fetchTree` part doesn’t know (nor has to know) that it’s fetching a flake.
But if the name gets defined in the flake (and used in the output path), then the `fetchTree` logic must be a bit more convoluted:
- It has to know whether it is fetching a flake
- When fetching a flake, it must get its `flake.nix`, evaluate it to find its `name`, and only then put it in the store with the correct `name`.

Reply:
Wouldn’t it be better to have this conversation outside the review widget? I feel like this aspect of the discussion is somewhat hidden & non-chronological.

Reply:
> * When fetching a flake, it must get its `flake.nix`, evaluate it to find its `name`, and only then put it in the store with the correct `name`.

This can be solved by adding `name` to the lock file.
We can use the same trick for other fetch-related parameters, like whether to fetch submodules or not, whether to git-crypt decrypt, etc, all of which would also be nice to configure in `flake.nix`.
`fetchTree` remains unaware of flake-ness.